ingest-app/src/cmr/ingest/validation/generic_document_validation.clj

```diff
@@ -0,0 +1,166 @@
+  "Provides functions to validate the ingest generic document"
+  (:require
+   [cheshire.core :as json]
+   [clojure.string :as string]
+   [cmr.common.cache :as cache]
+   [cmr.common.cache.in-memory-cache :as mem-cache]
+   [cmr.common.concepts :as concepts]
+   [cmr.common.generics :as generics]
+   [cmr.common.log :refer (info error)]
+   [cmr.common.services.errors :as errors]
+   [cmr.transmit.metadata-db :as mdb]))
+
+(def schema-validation-cache-key
+  "The cache key for the schema validation functions cache."
+  :schema-validation-functions)
+
+(def SCHEMA_CACHE_TIME
+  "The number of milliseconds schema validation functions will be cached."
+  (* 24 60 60 1000)) ;; 24 hours
+
+(defn create-schema-validation-cache
+  "Creates a cache for schema validation functions."
+  []
+  (mem-cache/create-in-memory-cache :ttl {} {:ttl SCHEMA_CACHE_TIME}))
+
+(defn- fetch-existing-concepts
+  "Fetch existing concepts of the given type from metadata-db for uniqueness validation"
+  [context concept-type]
+  (try
+    (mdb/find-concepts context {:latest true} concept-type)
+    (catch Exception e
+      (error (str "Error fetching concepts for generic validations: " (.getMessage e)))
+      []))) 
+
+(defn- validate-uniqueness
+  "Validates that the combination of field values is unique in the collection.
+   Returns a sequence of error messages if validation fails, empty sequence otherwise."
+  [context concept fields]
+  (let [concept-type (:concept-type concept)]
+    (if-let [existing-concepts (fetch-existing-concepts context concept-type)]
+      (let [;; Helper function to extract field values from a concept
+            get-field-values (fn [c fs]
+                               (let [metadata (json/parse-string (:metadata c) true)]
+                                 (mapv #(let [field-path (rest (string/split % #"\."))
+                                              keyword-path (mapv keyword field-path)]
+                                          (get-in metadata keyword-path))
+                                       fs)))
+
+            ;; Extract values for the specified fields from the current concept
+            field-values (get-field-values concept fields)
+
+            ;; Check if any other document in the collection has the same combination of values
+            duplicate-concepts (filter (fn [existing-concept]
+                                         (and (not= (:native-id existing-concept) (:native-id concept))
+                                              (not= (:deleted existing-concept) true)
+                                              (= (set (get-field-values existing-concept fields))
+                                                 (set field-values))))
+                                       existing-concepts)]
+        (if (seq duplicate-concepts)
+          (let [duplicate-concept-ids (map :concept-id duplicate-concepts)
+                field-names (mapv #(string/join "." (rest (string/split % #"\."))) fields)
```

Comment:
same question as above, can we use a common function for converting the jq style text to Clojure list